rfcs/0119-testing-conventions.md

```diff
@@ -0,0 +1,114 @@
+---
+feature: Defined conventions around testing of official Nixpkgs packages.
+start-date: 2021-09-03
+author: Jonathan Ringer
+co-authors:
+shepherd-team:
+shepherd-leader:
+related-issues:
+  - [RFC 0088 - Nixpkgs Breaking Change Policy](https://github.com/NixOS/rfcs/pull/88)
+---
+
+# Summary
+[summary]: #summary
+
+When updating or modifying packages, several conventions for testing regressions
+have been adopted. However, these practices are not standard, and generally it's not well
+defined how each testing method should be implemented. It would be beneficial to have
+an unambiguous way to say that a given package, and all downstream dependencies, have
+had as many automated test ran possible. This will give a high degree of certainty that
+a given change is less likely to manifest regressions once introduced on a release
+channel.
+
+Another desire of this rfc is also to have a way for various review tools
+(e.g. ofborg, hydra, nixpkgs-review) to have a standard way to determine if a
+package has additional tests which can help verify it's correctness.
+
+# Motivation
+[motivation]: #motivation
+
+Breakages are a constant painpoint for nixpkgs. It is a very poor user experience to
+have a configuration broken because one or more packages fail to build. Often when
+these breakages occur, it is because the change had a large impact on the entirity
+of nixpkgs; and unless there's a dedicated hydra jobset for the pull request, it's
+infeasible to expect pull request author(s) to verify every package that's affected
+by a change they are proposing. However, it is feasible to specify packages that
+are very likely to be affected by changes in another package, and use this information
+to help mitigate regressions from appearing in release channels.
+
+# Detailed design
+[design]: #detailed-design
+
+Standardize `passthru.tests.<name>` and `passthru.nixosTests.<name>` as a mechanism of 
+more expensive but automatic testing for nixpkgs. As well as encourage the usage of
+`checkPhase` or `installCheckPhase` when packaging within nixpkgs.
+
+Usage for `passthru.nixosTests.<name>`
+- Reserved for tests utilitizing the nixosTest utilties.
```

Comment:
This doesn't seem a meaningful distinction to me. Maybe we should flip this.

- Reserved for tests that are more resource intensive or that require additional system features such as kvm.
  - Generally these are tests using the `nixosTest` utilities.

Of course this makes the name less meaningful, but we can bikeshed that later.

Reply:
I was trying to keep it consistent in plurality with the already used `passthru.tests`. Also `pkgs.nixosTests` is how it's exposed in nixpkgs.